guillotina_amqp/job.py

```diff
@@ -94,42 +88,16 @@ def state_manager(self):
         return self._state_manager
 
     async def create_request(self):
-        req_data = self.data["req_data"]
-        url = req_data["url"]
-        parsed = urlparse(url)
-        dct = {
-            "method": req_data["method"],
-            "url": yarl.URL(url),
-            "path": parsed.path,
-            "headers": CIMultiDict(req_data["headers"]),
-            "raw_headers": tuple(
-                (k.encode("utf-8"), v.encode("utf-8"))
-                for k, v in req_data["headers"].items()
-            ),
-        }
-
-        message = self.base_request._message._replace(**dct)
-
-        payload_writer = mock.Mock()
-        payload_writer.write_eof.side_effect = noop
-        payload_writer.drain.side_effect = noop
-
-        protocol = mock.Mock()
-        protocol.transport = test_utils._create_transport(None)
-        protocol.writer = payload_writer
-
         request = self.base_request.__class__(
-            message,
-            EmptyPayload(),
-            protocol,
-            payload_writer,
-            self.task,
-            self.task._loop,
+            self.base_request.scheme,
+            self.base_request.method,
+            self.base_request.path,
+            self.base_request.query_string.encode("utf-8"),
+            self.base_request.raw_headers.copy(),
             client_max_size=self.base_request._client_max_size,
-            state=self.base_request._state.copy(),
         )
+        request._state = self.base_request._state.copy()
```

Comment:
Just had a quick chat with @lferran about the pattern that this code follows and all of the implications that it has, without having a cleary boundary between components where they have as unique coupling point a specific contract to meet, it's hard to scale the product in the future in terms of maintenance and all of the derivated side effects.

My take here would be if at any moment we have the excuse of changing how is this implemented I would consider change this pattern, but again by having a great excuse like for example having some teams taking the ownership of one of the Amqp consumers code.

However, there is this line of code [1] which is the one that smells most weird. May be would be worth it to review it.

[1] https://github.com/guillotinaweb/guillotina_amqp/pull/51/files#diff-cc5ae21c31bf6060f5c9b1a464010be3R74

Reply:
Can you describe more the problems maybe?

Reply:
with regards of the pattern of this specific line [1]? Maybe I'm wrong but both could be related somehow. meaning that we need to yes or yes reproduce a scenario with an open request for executing the function in a context where there is no even sense of HTTP protocol. As I said, I could be wrong.

Seems that the fact of having this magic wiring within the same application for having something being executed out of an HTTP server but designed for being executed using the same framework which is - unless I'm wrong - quite HTTP centric.

My point was that sometimes is better to be explicit, separate concerns, and be repetitive. Meaning that if you need to execute code in another context may be another way to do it would be by having as unique coupling point the explicit contract that you have for communicating both sides  - maybe a simple model, the producer and the consumer. This would allow us to provide freedom to both, without having to force patterns between each other.

I guess that sometimes is about choosing what works best at short term and sometimes is about choosing what would work best at the middle term.

As I said, I do not believe that changes would need to be done if they do not provide any value, but if and only if in the future we have a great excuse for changing this pattern maybe would be worth it to change it.

In any case, most likely I'm missing many things and details which they could be equally important!

[1] https://github.com/guillotinaweb/guillotina_amqp/blob/master/guillotina_amqp/job.py#L80

Reply:
gotcha, I agree. Maybe my comment here provides some context as well: https://github.com/guillotinaweb/guillotina_amqp/pull/51#discussion_r448936669